Data/SearchEngine/Autosuggest.hs

```diff
@@ -26,6 +29,51 @@ import qualified Data.IntSet as IntSet
 import qualified Data.Vector.Unboxed as Vec
 
 
+-- | Given an incomplete prefix query, find the set of documents that match
+-- possible completions of that query.  This should be less computationally
+-- expensive than 'queryAutosuggest' as it does not do any ranking of documents.
+-- However, it does not apply the pre-filter or post-filter limits, and the list
+-- may be large when the query terms occur in many documents.  The order of
+-- returned keys is unspecified.
+queryAutosuggestMatchingDocuments :: (Ix field, Bounded field, Ord key) =>
+                                     SearchEngine doc key field feature ->
+                                     [Term] -> Term -> [key]
+queryAutosuggestMatchingDocuments se@SearchEngine{searchIndex} precedingTerms partialTerm =
+    let (_, ds, _, _) = process se precedingTerms partialTerm
+    in map (SI.getDocKey searchIndex) (DocIdSet.toList ds)
+
+-- | Given an incomplete prefix query, return a predicate that indicates whether
+-- a key is in the set of documents that match possible completions of that
+-- query.  This is equivalent to calling 'queryAutosuggestMatchingDocuments' and
+-- testing whether the key is in the list, but should be more efficient.
+--
+-- This does not apply the pre-filter or post-filter limits.
+queryAutosuggestPredicate :: (Ix field, Bounded field, Ord key) =>
+                             SearchEngine doc key field feature ->
+                             [Term] -> Term -> (key -> Bool)
+queryAutosuggestPredicate se@SearchEngine{searchIndex} precedingTerms partialTerm =
+    let (_, ds, _, _) = process se precedingTerms partialTerm
+    in (\ key -> maybe False (flip DocIdSet.member ds) (SI.getKeyDocId searchIndex key))
+
+
+process :: (Ix field, Bounded field) =>
+           SearchEngine doc key field feature ->
+           [Term] -> Term -> ([TermId], DocIdSet, [(TermId, DocIdSet)], [TermId])
+process se precedingTerms partialTerm =
+   step_process
+   $ step_prep
+       precedingTerms partialTerm
+  where
+    -- Construct the auto-suggest query from the query terms
+    step_prep pre_ts t = mkAutosuggestQuery se pre_ts t
+
+    -- Find the appropriate subset of ts and ds
+    -- and an intermediate result that will be useful later:
+    -- { (t, ds ∩ ds_t) | t ∈ ts, ds ∩ ds_t ≠ ∅ }
+    step_process (ts, ds, pre_ts) = (ts', ds', tdss', pre_ts)
+      where
+        (tdss', ts', ds') = processAutosuggestQuery (ts, ds, pre_ts)
+
```

Comment:
Just a code order thing for readability: can we put these after `queryAutosuggest` within the file. I'd put `process` immediately after `queryAutosuggest` and then the two variants after that.